pennylane/optimize/spsa.py

```diff
@@ -0,0 +1,255 @@
+    """
+    # pylint: disable-msg=too-many-arguments
+    def __init__(self, maxiter=200, alpha=0.602, gamma=0.101, c=0.2, A=None, a=None):
+        if not A:
+            self.A = maxiter * 0.1
+        if not a:
+            self.a = 0.05 * (self.A + 1) ** alpha
+        self.c = c
+        self.alpha = alpha
+        self.gamma = gamma
+        self.k = 0
+        self.ak = self.a / (self.A + 1 + 1.0) ** self.alpha
+
+    def step_and_cost(self, objective_fn, *args, **kwargs):
+        """Update the parameter array :math:`x` with one step of the optimizer and return
+        the step and the corresponding objective function.
+
+        Args:
+            objective_fn (function): The objective function for optimization
+            *args : variable length argument array for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            tuple[list [array], float]: the new variable values :math:`x^{(t+1)}` and the
+            objective function output prior to the step.
+        """
+        g, forward = self.compute_grad(objective_fn, args, kwargs)
+        new_args = self.apply_grad(g, args)
+
+        if forward is None:
+            forward = objective_fn(*args, **kwargs)
+
+        # unwrap from list if one argument, cleaner return
+        if len(new_args) == 1:
+            return new_args[0], forward
+        return new_args, forward
+
+    def step(self, objective_fn, *args, **kwargs):
+        """Update trainable arguments with one step of the optimizer.
+
+        Args:
+            objective_fn (function): The objective function for optimization
+            *args : variable length argument array for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            list [array]: the new variable values :math:`x^{(t+1)}`.
+        """
+        g, _ = self.compute_grad(objective_fn, args, kwargs)
+        new_args = self.apply_grad(g, args)
+
+        # unwrap from list if one argument, cleaner return
+        if len(new_args) == 1:
+            return new_args[0]
+
+        return new_args
+
+    def increment_k(self):
+        """Increments k"""
+        self.k += 1
+
+    def compute_grad(self, objective_fn, args, kwargs):
+        r"""Compute approximation of gradient of the objective function at the
+        given point.
+
+        Args:
+            objective_fn (function): The objective function for optimization
+            args (tuple): tuple of NumPy array containing the current parameters
+                for objective function
+            kwargs (dict): keyword arguments for the objective function
+
+        Returns:
+            tuple (array): Numpy array containing the gradient
+                :math:`\hat{g}_k(\hat{\theta}_k)` and ``None``
+        """
+        self.increment_k()
+        ck = self.c / (self.k + 1.0) ** self.gamma
+        delta = list(args)
```

Comment:
Yes, it's needed later on to calculate the approximated gradient. I'd argue though, that it should be initialized to an empty list, rather than copying args and omit the initialization of grad to an empty list.